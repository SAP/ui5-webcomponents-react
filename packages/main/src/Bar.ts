import { deprecationNotice } from '@ui5/webcomponents-react-base';
import type { BarDomRef, BarPropTypes } from './webComponents/Bar';
import { Bar } from './webComponents/Bar';

deprecationNotice(
  'Bar',
  `importing from '@ui5/webcomponents-react/dist/Bar' is deprecated. Please use "import { Bar } from '@ui5/webcomponents-react';" instead.`
);

export { Bar };

export type { BarPropTypes, BarDomRef };
