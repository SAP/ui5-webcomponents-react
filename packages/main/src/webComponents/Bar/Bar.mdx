import { ControlsWithNote, DocsHeader, Footer } from '@sb/components';
import { Canvas, Meta } from '@storybook/blocks';
import * as ComponentStories from './Bar.stories';

<Meta of={ComponentStories} />

<DocsHeader since="0.13.0" />

<br />

## Example

<Canvas of={ComponentStories.Default} />

## Properties

<ControlsWithNote of={ComponentStories.Default} />

# More Examples

<br />

## Bar with custom elements

<Canvas of={ComponentStories.WithCustomElements} />

<Footer />
