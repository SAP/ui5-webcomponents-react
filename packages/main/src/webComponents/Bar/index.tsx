import '@ui5/webcomponents-fiori/dist/Bar.js';
import { ReactNode } from 'react';
import { BarDesign } from '../../enums';
import { CommonProps } from '../../interfaces/CommonProps';
import { Ui5DomRef } from '../../interfaces/Ui5DomRef';
import { withWebComponent } from '../../internal/withWebComponent';
import { UI5WCSlotsNode } from '../../types';

interface BarAttributes {
  /**
   * Defines the component's design.
   *
   * **Note:** Available options are:
   *
   * *   `Header`
   * *   `Subheader`
   * *   `Footer`
   * *   `FloatingFooter`
   */
  design?: BarDesign | keyof typeof BarDesign;
}

export interface BarDomRef extends BarAttributes, Ui5DomRef {}

export interface BarPropTypes extends BarAttributes, CommonProps {
  /**
   * Defines the content in the middle of the bar
   */
  children?: ReactNode | ReactNode[];
  /**
   * Defines the content at the end of the bar
   *
   * __Note:__ This prop will be rendered as [slot](https://www.w3schools.com/tags/tag_slot.asp) (`slot="endContent"`).
   * Since you can't change the DOM order of slots when declaring them within a prop, it might prove beneficial to manually mount them in the body of the component, especially when facing problems with the reading order of screen readers.
   *
   * __Note:__ When passing a custom React component to this prop, you have to make sure your component reads the `slot` prop and appends it to the most outer element of your component.
   * Learn more about it [here](https://sap.github.io/ui5-webcomponents-react/?path=/docs/knowledge-base-handling-slots--page).
   */
  endContent?: UI5WCSlotsNode | UI5WCSlotsNode[];
  /**
   * Defines the content at the start of the bar
   *
   * __Note:__ This prop will be rendered as [slot](https://www.w3schools.com/tags/tag_slot.asp) (`slot="startContent"`).
   * Since you can't change the DOM order of slots when declaring them within a prop, it might prove beneficial to manually mount them in the body of the component, especially when facing problems with the reading order of screen readers.
   *
   * __Note:__ When passing a custom React component to this prop, you have to make sure your component reads the `slot` prop and appends it to the most outer element of your component.
   * Learn more about it [here](https://sap.github.io/ui5-webcomponents-react/?path=/docs/knowledge-base-handling-slots--page).
   */
  startContent?: UI5WCSlotsNode | UI5WCSlotsNode[];
}

/**
 * The Bar is a container which is primarily used to hold titles, buttons and input elements and its design and functionality is the basis for page headers and footers. The component consists of three areas to hold its content - startContent slot, default slot and endContent slot. It has the capability to center content, such as a title, while having other components on the left and right side.
 *
 * __Note:__ This component is a web component developed by the UI5 Web Componentsâ€™ team.
 *
 * <ui5-link href="https://sap.github.io/ui5-webcomponents/playground/components/Bar" target="_blank">UI5 Web Components Playground</ui5-link>
 */
const Bar = withWebComponent<BarPropTypes, BarDomRef>('ui5-bar', ['design'], [], ['endContent', 'startContent'], []);

Bar.displayName = 'Bar';

Bar.defaultProps = {
  design: BarDesign.Header
};

export { Bar };
