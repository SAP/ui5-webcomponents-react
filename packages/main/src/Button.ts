import { deprecationNotice } from '@ui5/webcomponents-react-base';
import type { ButtonDomRef, ButtonPropTypes } from './webComponents/Button';
import { Button } from './webComponents/Button';

deprecationNotice(
  'Button',
  `importing from '@ui5/webcomponents-react/dist/Button' is deprecated. Please use "import { Button } from '@ui5/webcomponents-react';" instead.`
);

export { Button };

export type { ButtonPropTypes, ButtonDomRef };
