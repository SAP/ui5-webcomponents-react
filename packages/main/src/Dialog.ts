import { deprecationNotice } from '@ui5/webcomponents-react-base';
import type { DialogDomRef, DialogPropTypes } from './webComponents/Dialog';
import { Dialog } from './webComponents/Dialog';

deprecationNotice(
  'Dialog',
  `importing from '@ui5/webcomponents-react/dist/Dialog' is deprecated. Please use "import { Dialog } from '@ui5/webcomponents-react';" instead.`
);

export { Dialog };

export type { DialogPropTypes, DialogDomRef };
