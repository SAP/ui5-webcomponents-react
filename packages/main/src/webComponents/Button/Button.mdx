import { Canvas, Markdown, Meta } from '@storybook/blocks';
import { ControlsWithNote, DocsHeader, DomRefTable, Footer } from '@sb/components';
import Description from './ButtonDescription.md?raw';
import ButtonDomRef from './ButtonDomRef.json';
import * as ComponentStories from './Button.stories';

<Meta of={ComponentStories} />

<DocsHeader />

<br />

## Example

<Canvas of={ComponentStories.Default} />

## Properties

<ControlsWithNote of={ComponentStories.Default} />

<Markdown>{Description}</Markdown>

<DomRefTable rows={ButtonDomRef} />

<Footer />
