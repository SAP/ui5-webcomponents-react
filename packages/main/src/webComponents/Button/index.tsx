import { ButtonDesign } from '@ui5/webcomponents-react/dist/ButtonDesign';
import { withWebComponent } from '@ui5/webcomponents-react/dist/withWebComponent';
import { CommonProps } from '@ui5/webcomponents-react/interfaces/CommonProps';
import { ReactNode, MouseEventHandler } from 'react';

import '@ui5/webcomponents/dist/Button';

export interface ButtonPropTypes extends Omit<CommonProps, 'onClick'> {
  /**
   * Sets the accessible aria name of the component.
   */
  accessibleName?: string;
  /**
   * Defines the component design.
   *
   * **Note:**
   *
   * *   `Default`
   * *   `Emphasized`
   * *   `Positive`
   * *   `Negative`
   * *   `Transparent`
   * *   `Attention`
   */
  design?: ButtonDesign | keyof typeof ButtonDesign;
  /**
   * Defines whether the component is disabled (default is set to `false`). A disabled component can't be pressed or focused, and it is not in the tab chain.
   */
  disabled?: boolean;
  /**
   * Defines the icon to be displayed as graphical element within the component. The SAP-icons font provides numerous options.
   *
   * Example: See all the available icons in the <ui5-link target="_blank" href="https://openui5.hana.ondemand.com/test-resources/sap/m/demokit/iconExplorer/webapp/index.html" class="api-table-content-cell-link">Icon Explorer</ui5-link>.
   */
  icon?: string;
  /**
   * Defines whether the icon should be displayed after the component text.
   */
  iconEnd?: boolean;
  /**
   * When set to `true`, the component will automatically submit the nearest form element upon `press`.
   *
   * **Important:** For the `submits` property to have effect, you must add the following import to your project: `import "@ui5/webcomponents/dist/features/InputElementsFormSupport.js";`
   */
  submits?: boolean;
  /**
   * Defines the text of the component.
   *
   * **Note:** Although this slot accepts HTML Elements, it is strongly recommended that you only use text in order to preserve the intended design.
   */
  children?: ReactNode | ReactNode[];
  /**
   * Fired when the component is activated either with a mouse/tap or by using the Enter or Space key.
   *
   * **Note:** The event will not be fired if the `disabled` property is set to `true`.
   */
  onClick?: MouseEventHandler<HTMLElement>;
}

/**
 * The `Button` component represents a simple push button. It enables users to trigger actions by clicking or tapping the `Button`, or by pressing certain keyboard keys, such as Enter.
 *
 * <ui5-link href="https://sap.github.io/ui5-webcomponents/playground/components/Button" target="_blank">UI5 Web Components Playground</ui5-link>
 */
const Button = withWebComponent<ButtonPropTypes>(
  'ui5-button',
  ['accessibleName', 'design', 'icon'],
  ['disabled', 'iconEnd', 'submits'],
  [],
  ['click']
);

Button.displayName = 'Button';

Button.defaultProps = {
  design: ButtonDesign.Default,
  disabled: false,
  iconEnd: false,
  submits: false
};

export { Button };
