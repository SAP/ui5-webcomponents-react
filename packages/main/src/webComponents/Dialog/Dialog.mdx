import { ControlsWithNote, DocsHeader, DomRefTable, Footer } from '@sb/components';
import { Canvas, Markdown, Meta } from '@storybook/blocks';
import * as ComponentStories from './Dialog.stories';
import Description from './DialogDescription.md?raw';
import DialogDomRef from './DialogDomRef.json';

<Meta of={ComponentStories} />

<DocsHeader />

<br />

## Example

<Canvas of={ComponentStories.Default} />

## Properties

<ControlsWithNote of={ComponentStories.Default} />

<Markdown>{Description}</Markdown>

<DomRefTable rows={DialogDomRef} />

## Opening Dialogs

<br />

#### since 0.22.0

We recommend opening and closing the `Dialog` component in a declarative way using the `open` prop.
You can still use the imperative way which is outlined below.

```jsx
const MyComponentWithDialog = () => {
  const [dialogIsOpen, setDialogIsOpen] = useState(false);
  return (
    <>
      <Button
        onClick={() => {
          setDialogIsOpen(true);
        }}
      >
        Open Dialog
      </Button>
      <Dialog open={dialogIsOpen} />
    </>
  );
};
```

#### before 0.22.0

`Dialogs` can be opened by attaching a `ref` to the component and then calling the corresponding **`show`** method.

```jsx
const DialogComponent = () => {
  const dialogRef = useRef(null);
  const onButtonClick = () => {
    dialogRef.current.show();
  };
  return (
    <>
      <Button onClick={onButtonClick}>Open Dialog</Button>
      <Dialog ref={dialogRef}>Some Content</Dialog>
    </>
  );
};
```

## Using Dialogs inside other components

`Dialogs` are often used within other components, when opened this could sometimes have unwanted side effects.
In this case, we recommend using [createPortal](https://reactjs.org/docs/portals.html) to mount the `Dialog` outside of the DOM hierarchy of the parent component.

```jsx
const DialogComponent = () => {
  const dialogRef = useRef(null);
  const onButtonClick = () => {
    dialogRef.current.show();
  };
  return (
    <>
      <Button onClick={onButtonClick}>Open Dialog</Button>
      {createPortal(<Dialog ref={dialogRef} />, document.body)}
    </>
  );
};
```

## Closing Dialogs

Closing `Dialogs` works in the same way as opening them.
You can either set the `open` prop to `false` or attaching a `ref` on which the corresponding `close` method is called.

```jsx
const DialogComponent = () => {
  const dialogRef = useRef(null);
  const handleOpen = (e) => {
    dialogRef.current.show();
  };
  const handleClose = () => {
    dialogRef.current.close();
  };
  return (
    <>
      <Button onClick={handleOpen}>Open Dialog</Button>
      <Dialog ref={dialogRef}>
        <Button onClick={handleClose}>Close Dialog</Button>
      </Dialog>
    </>
  );
};
```

# More Examples

## Dialog with content

<Canvas of={ComponentStories.WithContent} />

<Footer />
