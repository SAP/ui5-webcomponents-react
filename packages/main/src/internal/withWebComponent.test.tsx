import { render, fireEvent, screen } from '@shared/tests';
import { Button } from '@ui5/webcomponents-react/dist/Button';
import { Dialog } from '@ui5/webcomponents-react/dist/Dialog';
import { Bar } from '@ui5/webcomponents-react/dist/Bar';
import {
  setCustomElementsScopingSuffix,
  setCustomElementsScopingRules
} from '@ui5/webcomponents-base/dist/CustomElementsScope.js';
import React from 'react';

describe('withWebComponent', () => {
  test('Unmount Event Handlers correctly after prop update', () => {
    const callback = jest.fn();

    const { rerender, unmount } = render(<Button onClick={callback}>Click Me</Button>);
    fireEvent.click(screen.getByText('Click Me'));
    expect(callback).toHaveBeenCalledTimes(1);

    const newCallback = jest.fn();
    rerender(<Button onClick={newCallback}>Click Me</Button>);
    fireEvent.click(screen.getByText('Click Me'));
    expect(callback).toHaveBeenCalledTimes(1);
    expect(newCallback).toHaveBeenCalledTimes(1);

    fireEvent.click(screen.getByText('Click Me'));
    expect(newCallback).toHaveBeenCalledTimes(2);

    rerender(<Button>Click Me</Button>);
    fireEvent.click(screen.getByText('Click Me'));
    expect(callback).toHaveBeenCalledTimes(1);
    expect(newCallback).toHaveBeenCalledTimes(2);

    unmount();
  });

  test('Slots accept both Fragment and regular nodes', () => {
    const callbackFooter1 = jest.fn();
    const callbackFooter2 = jest.fn();

    const { rerender } = render(
      <Dialog footer={<Button onClick={callbackFooter1}>Footer Button 1</Button>}>My Dialog Content</Dialog>
    );

    fireEvent.click(screen.getByText('Footer Button 1'));
    expect(callbackFooter1).toHaveBeenCalledTimes(1);

    rerender(
      <Dialog
        footer={
          <>
            <Button onClick={callbackFooter1}>Footer Button 1</Button>
            <Button onClick={callbackFooter2}>Footer Button 2</Button>
          </>
        }
      >
        My Dialog Content
      </Dialog>
    );

    fireEvent.click(screen.getByText('Footer Button 1'));
    expect(callbackFooter1).toHaveBeenCalledTimes(2);

    fireEvent.click(screen.getByText('Footer Button 2'));
    expect(callbackFooter2).toHaveBeenCalledTimes(1);
  });

  test('correct handling of boolean props', () => {
    const { rerender } = render(<Button disabled>Click Me</Button>);

    expect(screen.getByText('Click Me')).toHaveAttribute('disabled');

    rerender(<Button>Click Me</Button>);
    expect(screen.getByText('Click Me')).not.toHaveAttribute('disabled');
  });

  test('conditional rendering', () => {
    const { asFragment } = render(
      <Bar
        startContent={
          <>
            {false && <span>I'm not here</span>}
            <span>I'm here!</span>
          </>
        }
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('destruct fragments in slots', () => {
    const { asFragment } = render(
      <Bar
        startContent={
          <>
            {false && <span>I'm not here</span>}
            <span>I'm here!</span>
            <>
              {false && <span>I'm not here</span>}
              <span>I'm here nested level 1!</span>
              <>
                {false && <span>I'm not here</span>}
                <span>I'm here nested level 2!</span>
                <>
                  <span>I'm the only child of a fragment in level 3!</span>
                </>
              </>
            </>
            <>
              <span>I'm the only child of a fragment in level 1!</span>
            </>
            <>{false && <span>I'm an empty fragment</span>}</>
            <>
              <span>I'm the only child of a fragment in level 1!</span>
              {[<span>I'm in an array inside of a fragment!</span>, <span>Me too!</span>]}
            </>
          </>
        }
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('scoping', () => {
    setCustomElementsScopingSuffix('ui5-wcr');
    const { asFragment, rerender } = render(<Button>Scoping Test</Button>);
    expect(asFragment()).toMatchSnapshot();

    // now exclude the button
    setCustomElementsScopingRules({ include: [/^ui5-/], exclude: [/^ui5-button/] });
    rerender(<Button>Scoping Test</Button>);
    expect(asFragment()).toMatchSnapshot();
  });
});
