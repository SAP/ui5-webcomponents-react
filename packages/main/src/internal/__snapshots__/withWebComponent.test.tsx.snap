// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withWebComponent conditional rendering 1`] = `
<DocumentFragment>
  <ui5-bar
    data-sap-ui-fastnavgroup="true"
    design="Header"
    ui5-bar=""
  >
    <span
      slot="startContent"
    >
      I'm here!
    </span>
  </ui5-bar>
</DocumentFragment>
`;

exports[`withWebComponent destruct fragments in slots 1`] = `
<DocumentFragment>
  <ui5-bar
    data-sap-ui-fastnavgroup="true"
    design="Header"
    ui5-bar=""
  >
    <span
      slot="startContent"
    >
      I'm here!
    </span>
    <span
      slot="startContent"
    >
      I'm here nested level 1!
    </span>
    <span
      slot="startContent"
    >
      I'm here nested level 2!
    </span>
    <span
      slot="startContent"
    >
      I'm the only child of a fragment in level 3!
    </span>
    <span
      slot="startContent"
    >
      I'm the only child of a fragment in level 1!
    </span>
    <span
      slot="startContent"
    >
      I'm the only child of a fragment in level 1!
    </span>
    <span
      slot="startContent"
    >
      I'm in an array inside of a fragment!
    </span>
    <span
      slot="startContent"
    >
      Me too!
    </span>
  </ui5-bar>
</DocumentFragment>
`;

exports[`withWebComponent scoping 1`] = `
<DocumentFragment>
  <ui5-button-ui5-wcr
    design="Default"
  >
    Scoping Test
  </ui5-button-ui5-wcr>
</DocumentFragment>
`;

exports[`withWebComponent scoping 2`] = `
<DocumentFragment>
  <ui5-button
    design="Default"
    ui5-button=""
  >
    Scoping Test
  </ui5-button>
</DocumentFragment>
`;
